### PR TITLE
`reactpy.run` and `configure(...)` refactoring

### PR DESCRIPTION
**Pull request by Archmonger**
## Issues

Right now `reactpy.run` can fail in fairly unpredictable ways which give little in terms of stack traces. 

## Summary

This refactors the `reactpy.run` call stack in order allow failures that are user-readable.

- Change `reactpy.backends.utils.find_all_implementations()` to first try to import `<backend_name>` before importing `reactpy.backend.<backend_name>`
   - Allows for missing sub-dependencies to not cause `reactpy.run` to silently fail
- Import `uvicorn` directly within `serve_with_uvicorn` in order to defer import.
   - Allows for `ModuleNotFound: Could not import uvicorn` exception to tell the user what went wrong
- Added `CommonOptions.serve_index_route: bool`
   - Allows us to not clutter the route patterns when it's not needed
   - There are real circumstances where a user might want the index route to 404
- Fix bug where in-use ports are being assigned on Windows.
   - Removes `allow_reuse_waiting_ports` parameter on `find_available_port()` 
- Rename `BackendImplementation` to `BackendProtocol`
- Change load order of `SUPPORTED_PACKAGES` so that `FastAPI` has a chance to run before `starlette`
- Rename `SUPPORTED_PACKAGES` to `SUPPORTED_BACKENDS`
- Refactor `reactpy.backend.*` code to be more human readable
- Use f-strings where possible
- Merge `if` statements where possible
- Use `contextlib.supress` where possible
- Remove defunct `requirements.txt` file 

## Checklist

- [ ] Tests have been included for all bug fixes or added functionality.
- [x] The `changelog.rst` has been updated with any significant changes.

### CODE REVIEW

**Comment by Archmonger** [comment]
@rmorshea `npm install -g npm@8.5.0` seems to have broken in the github workflow in the last 24 hours. Not sure why, but I tried bumping `actions/setup-node` version. Did not fix it.

**Comment by rmorshea** [comment]
I was planning on doing a 1.0.1 release tomorrow. Should this wait for the next release?

**Comment by Archmonger** [comment]
Yes, technically this release has an API change (`BackendImplementation` to `BackendProtocol`), so we shouldn't put this in a hotfix release.

**Comment by Archmonger** [comment]
I discovered that our `flask` backend doesn't work with `uvicorn` due to `WsgiToAsgi` not allowing [non-HTTP scopes](https://github.com/django/asgiref/blob/78128f185b17b5f23e269bfb052d9d0ffb05c05b/asgiref/wsgi.py#L35). When removing this restriction, this showed that  `flask-sock` relies on via non-standard WSGI extensions which `WsgiToAsgi` does not support.

I attempted to rewrite `WsgiToAsgi` locally but it broke a ton of other stuff.

Ref: https://github.com/miguelgrinberg/flask-sock/issues/58

File: src/py/reactpy/reactpy/backend/_common.py
Comment (rmorshea) [inline src/py/reactpy/reactpy/backend/_common.py]: You can put the uvicorn import in an `if TYPE_CHECKING` block instead of making this `Any`.

File: src/py/reactpy/reactpy/backend/default.py
Comment (rmorshea) [inline src/py/reactpy/reactpy/backend/default.py]:
```suggestion
    def __init__(self, *args: Any, **kwds: Any) -> NoReturn:
```

File: src/py/reactpy/reactpy/backend/types.py
Comment (rmorshea) [inline src/py/reactpy/reactpy/backend/types.py]: While I doubt changing this name will break anything for people we may want to avoid unnecessary renames.

If we do want to rename this, while I haven't been perfect, I've generally named protocols that are object interfaces with a `Type` suffix and those that are callables `Func`. So by that convention we'd name this `BackendType`.

File: src/py/reactpy/reactpy/backend/utils.py
Comment (rmorshea) [inline src/py/reactpy/reactpy/backend/utils.py]: Unclear why this is better.

File: src/py/reactpy/reactpy/backend/utils.py
Comment (rmorshea) [inline src/py/reactpy/reactpy/backend/utils.py:46]: There's probably a bunch of other places where I used f-string in logs

**Review by rmorshea — COMMENTED** [review]
Generally just minor things to correct.

File: src/py/reactpy/reactpy/backend/default.py
Comment (rmorshea) [reply src/py/reactpy/reactpy/backend/default.py]: The docstring should be updated to reflect that it should not be used too.

File: src/py/reactpy/reactpy/backend/types.py
Comment (Archmonger) [reply src/py/reactpy/reactpy/backend/types.py]: Given that `BackendType` is going to have at minimum 4 attributes, and possibly more in the future, should we consider changing our backends to use classes instead?

It feels really disjointed to read right now.

File: src/py/reactpy/reactpy/backend/utils.py
Comment (Archmonger) [reply src/py/reactpy/reactpy/backend/utils.py:46]: Yup, I didn't want to clutter this PR too much though. I think covering other files should be a follow up PR.

File: src/py/reactpy/reactpy/backend/default.py
Comment (Archmonger) [reply src/py/reactpy/reactpy/backend/default.py]: A docstring that `__init__` should not be used, or am I misunderstanding?

Comment (rmorshea) [reply src/py/reactpy/reactpy/backend/default.py]: Yeah, the default options should never be constructed.

File: src/py/reactpy/reactpy/backend/types.py
Comment (rmorshea) [reply src/py/reactpy/reactpy/backend/types.py]: Given that none of the methods require any shared state (and I'd like to avoid that if possible) I don't think we need these to be attached to classes.

File: src/py/reactpy/reactpy/backend/tornado.py
Comment (rmorshea) [inline src/py/reactpy/reactpy/backend/tornado.py]: These comments need to be updated with the new name.

File: src/py/reactpy/reactpy/backend/types.py
Comment (Archmonger) [reply src/py/reactpy/reactpy/backend/types.py]: Conceptually they're still tied together though. They could all be static methods in a class perhaps?

When I was looking at creating a backend, it was fairly hideous to stare at our existing files to figure out what's going on.

Comment (rmorshea) [reply src/py/reactpy/reactpy/backend/types.py]: If things become more complicated we can consider doing that. For now, it seems like it would be similarly complex to make that change without breaking the ability to do `from reactpy.backend.package import configure`.

File: src/py/reactpy/reactpy/backend/default.py
Comment (rmorshea) [inline src/py/reactpy/reactpy/backend/default.py]: I don't think there's any reason someone should subclass this

```suggestion
# BackendProtocol.Options
class Options:  # nocov
    """Do not instantiate - default implementation has no options"""

    def __init__(self, *args: Any, **kwds: Any) -> NoReturn:
        msg = "Default implementation has no options."
        raise ValueError(msg)
```

File: src/py/reactpy/reactpy/backend/utils.py
Comment (rmorshea) [inline src/py/reactpy/reactpy/backend/utils.py]: Not sure why we're deleting this. Without this, if you run the stop and then immediately restart the server, it's very likely to use a different address. This is rather annoying since you can't simply return to your existing browser tab and hit refresh.

Comment (Archmonger) [reply src/py/reactpy/reactpy/backend/utils.py]: Reason is mentioned in the PRs description.

Fixes a bug where in-use ports are being assigned on Windows. Deleting it didn't seem to impact Linux.

Comment (rmorshea) [reply src/py/reactpy/reactpy/backend/utils.py]: Doesn't break on Linux, but every time you restart the server you'll get a different port. Perhaps we can just vary the behavior by platform.

Comment (rmorshea) [reply src/py/reactpy/reactpy/backend/utils.py]: I should double check that, but that is why I recall introducing that.

Comment (Archmonger) [reply src/py/reactpy/reactpy/backend/utils.py]: Okay let me know, if that's behavior on Linux I can just change the default when running win32.

**Comment by Archmonger** [comment]
@rmorshea This seems to be mergeable, but needs approval.

**Review by rmorshea — APPROVED** [review]

**Pull request merged**
